src/openms/source/CHEMISTRY/DecoyGenerator.cpp

```diff
@@ -35,86 +35,112 @@
 #include <OpenMS/CHEMISTRY/AASequence.h>
 #include <OpenMS/CHEMISTRY/ProteaseDigestion.h>
 #include <OpenMS/CHEMISTRY/DecoyGenerator.h>
+#include <OpenMS/CONCEPT/Macros.h>
 
 #include <random>
 
 using namespace OpenMS;
 
 // static
-AASequence DecoyGenerator::reverseProtein(const AASequence& aas)
+AASequence DecoyGenerator::reverseProtein(const AASequence& protein)
 {
-    String s = aas.toUnmodifiedString();
-    std::reverse(s.begin(), s.end());
-    return AASequence::fromString(s);
+  OPENMS_PRECONDITION(!protein.isModified(), "Decoy generation only supports unmodified proteins.");
+  String s = protein.toUnmodifiedString();
+  std::reverse(s.begin(), s.end());
+  return AASequence::fromString(s);
 }
 
 // static
-AASequence DecoyGenerator::reversePeptides(const AASequence& aas, const String& protease)
+AASequence DecoyGenerator::reversePeptides(const AASequence& protein, const String& protease)
 {
-    std::vector<AASequence> peptides;
-    ProteaseDigestion ed;
-    ed.setMissedCleavages(0); // important as we want to reverse between all cutting sites
-    ed.setEnzyme(protease);
-    ed.setSpecificity(EnzymaticDigestion::SPEC_NONE);
-    ed.digest(aas, peptides);    
-    String pseudo_reversed;
-    for (const auto & aas : peptides)
-    {
-        std::string s = aas.toUnmodifiedString();
-        auto last = --s.end(); // don't reverse enzymatic cutting site
-        std::reverse(s.begin(), last);
-        pseudo_reversed += s;
-    }
-    return AASequence::fromString(pseudo_reversed);
+  OPENMS_PRECONDITION(!protein.isModified(), "Decoy generation only supports unmodified proteins.");
+  std::vector<AASequence> peptides;
+  ProteaseDigestion ed;
+  ed.setMissedCleavages(0); // important as we want to reverse between all cutting sites
+  ed.setEnzyme(protease);
+  ed.setSpecificity(EnzymaticDigestion::SPEC_NONE);
+  ed.digest(protein, peptides);    
+  String pseudo_reversed;
+  for (int i = 0; i < peptides.size() - 1; ++i)
+  {
+    std::string s = peptides[i].toUnmodifiedString();
+    auto last = --s.end(); // don't reverse enzymatic cutting site
+    std::reverse(s.begin(), last);
+    pseudo_reversed += s;
+  }
+  // the last peptide of a protein is not an enzymatic cutting site so we do a full reverse
+  std::string s = peptides[peptides.size() - 1 ].toUnmodifiedString();
+  std::reverse(s.begin(), s.end());
+  pseudo_reversed += s;
+  return AASequence::fromString(pseudo_reversed);
 }
 
-AASequence DecoyGenerator::shufflePeptide(
+AASequence DecoyGenerator::shufflePeptides(
         const AASequence& protein,
         const String& protease,
         const int max_attempts,
         int seed)
 {
-    std::mt19937 rng(seed);
+  OPENMS_PRECONDITION(!protein.isModified(), "Decoy generation only supports unmodified proteins.");
+  std::mt19937 rng(seed);
```

Comment:
hmm seems to lead to different results on windows. Any suggestions on how to make sure the tests have identical results on all platforms?

Reply:
Seems like you have to write your own shuffle. the twister should be platform-independent.

Reply:
yep seems so... would have been nice to let users choose if they need the X% faster implementation or something that is platform-independent 
https://stackoverflow.com/questions/51929085/cross-platform-random-reproducibility

Reply:
@jpfeuffer @cbielow where should I put a "portable_shuffle" somewhere in Math or Concept?

Reply:
or just use https://www.boost.org/doc/libs/1_71_0/libs/range/doc/html/range/reference/algorithms/mutating/random_shuffle.html ?